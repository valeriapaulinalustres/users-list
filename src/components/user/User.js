import React, { useRef } from 'react';
import './user.css';
import Swal from 'sweetalert2';

function User({ id, email, name, phone, deleteUser, edit, updateUsers, onEnter, onLeave, validateName, 
  validateMail, invalidName, invalidPhone, invalidEmail }) {

  const inputNameRef = useRef()
  const inputMailRef = useRef()
  const inputPhoneRef = useRef()

  //function for deleting users
  const handleDelete = () => {
    Swal.fire({
      title: 'Do you want to delete the user?',
      showDenyButton: true,
      showCancelButton: true,
      confirmButtonText: 'Yes',
      denyButtonText: `No`,
      confirmButtonColor: "#8dec8d",
    }).then((result) => {
      if (result.isConfirmed) {
        deleteUser(id)
        Swal.fire('User deleted')
      } else if (result.isDenied) {
        Swal.fire('User hasnÂ´t been deleted')
      }
    })
  }

  //function for saving changes in users
  const handleSave = () => {
    
    const validName = validateName(inputNameRef.current.value);
    if (!validName) return invalidName();

    if (inputPhoneRef.current.value == "") return invalidPhone();

    const validEmail = validateMail(inputMailRef.current.value);
    if (!validEmail) return invalidEmail();

    Swal.fire({
      title: 'Do you want to save the changes?',
      showDenyButton: true,
      showCancelButton: true,
      confirmButtonText: 'Yes',
      denyButtonText: `No`,
      confirmButtonColor: "#8dec8d"
    }).then((result) => {
      if (result.isConfirmed) {
        updateUsers(inputNameRef.current.value, inputPhoneRef.current.value, inputMailRef.current.value, id)
        Swal.fire('Changes has been saved')
      } else if (result.isDenied) {
        Swal.fire('Changes has not been saved')
      }
    })
  }

  //function for rendering edit mode
  const renderEdit = () => {
    return <div className='input-edit-container'>
      <input
        type="text"
        defaultValue={name}
        ref={inputNameRef}
        id={id}
        className="input"
      />
      <input
        type="text"
        defaultValue={phone}
        ref={inputPhoneRef}
        id={id}
        className="input"
      />
      <input
        type="text"
        defaultValue={email}
        ref={inputMailRef}
        id={id}
        className="input"
      />
      <button onClick={handleSave} className="button" onMouseEnter={onEnter} onMouseLeave={onLeave}>Save</button>
    </div>
  }

  //function for rendering users list mode
  const renderUser = () => {
    return <div className='card' role="userContainer">
      <div className='card-header'>
        <h2 role="userName">{name}</h2>
        <button onClick={handleDelete} className="button" onMouseEnter={onEnter} onMouseLeave={onLeave}>Delete</button>
      </div>
      <h3>phone: {phone}</h3>
      <h3>email: {email}</h3>
    </div>
  }

  return (
    <div>
      {edit
        ? renderEdit()
        : renderUser()
      }
    </div>
  )
}

export default User