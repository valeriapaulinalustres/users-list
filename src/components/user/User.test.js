import { render, screen } from '@testing-library/react';
import User from './User';


const nameProp = "Leanne Graham"
  
  it("matches snapshot (name PROP)", () => {
    const { asFragment } = render(<User prop={nameProp} />);
    expect(asFragment()).toMatchSnapshot();
  });

  const phoneProp = "1-770-736-8031 x56442"
  
  it("matches snapshot (phone PROP)", () => {
    const { asFragment } = render(<User prop={phoneProp} />);
    expect(asFragment()).toMatchSnapshot();
  });

  const mailProp = "Sincere@april.biz"
  
  it("matches snapshot (mail PROP)", () => {
    const { asFragment } = render(<User prop={mailProp} />);
    expect(asFragment()).toMatchSnapshot();
  });



test('should render user Container', async () => {
  render(<User />)
    const userContainer = await screen.findByRole('userContainer');
    expect(userContainer).toBeInTheDocument();
  });

  test('should render a person name', async () => {
    render(<User />)
    const userName = await screen.findByRole(/userName/i)
    expect(userName).toBeInTheDocument();
  });

  test('should render a new card passed as prop', async () => {
    render(<User />)
    const userName = await screen.findByRole(/userName/i)
    expect(userName).toBeInTheDocument();
  });