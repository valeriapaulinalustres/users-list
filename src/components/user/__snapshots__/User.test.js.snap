// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot (mail PROP) 1`] = `
<DocumentFragment>
  <div>
    <div
      class="card"
      role="userContainer"
    >
      <div
        class="card-header"
      >
        <h2
          role="userName"
        />
        <button
          class="button"
        >
          Delete
        </button>
      </div>
      <h3>
        phone: 
      </h3>
      <h3>
        email: 
      </h3>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot (name PROP) 1`] = `
<DocumentFragment>
  <div>
    <div
      class="card"
      role="userContainer"
    >
      <div
        class="card-header"
      >
        <h2
          role="userName"
        />
        <button
          class="button"
        >
          Delete
        </button>
      </div>
      <h3>
        phone: 
      </h3>
      <h3>
        email: 
      </h3>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot (phone PROP) 1`] = `
<DocumentFragment>
  <div>
    <div
      class="card"
      role="userContainer"
    >
      <div
        class="card-header"
      >
        <h2
          role="userName"
        />
        <button
          class="button"
        >
          Delete
        </button>
      </div>
      <h3>
        phone: 
      </h3>
      <h3>
        email: 
      </h3>
    </div>
  </div>
</DocumentFragment>
`;
